File: 543. Diameter of Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        if(root == NULL)return 0;
        int lh, rh;
        int ld = getHeightAndDiameter(root->left, lh);
        int rd = getHeightAndDiameter(root->right, rh);
        return max(lh + rh + 2, max(ld, rd));
    }
    // return diameter, use reference to return h
    int getHeightAndDiameter(TreeNode* root, int& h) {
        if(root == NULL) {
            h = -1;
            return 0;
        }
        int lh, rh;
        int ld = getHeightAndDiameter(root->left, lh);
        int rd = getHeightAndDiameter(root->right, rh);

        h = max(lh, rh) + 1;

        return max(lh + rh + 2, max(ld, rd));
    }

};

File: 543. Diameter of Binary Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int diameterOfBinaryTree(TreeNode root) {
        if(root == null)return 0;

        int cross = depth(root.left) + depth(root.right) + 2;
        int left = diameterOfBinaryTree(root.left);
        int right = diameterOfBinaryTree(root.right);

        return Math.max(cross, Math.max(left, right));
    }

    public int depth(TreeNode root) {
        if(root == null)return -1;
        return Math.max(depth(root.left), depth(root.right)) + 1;
    }
}
